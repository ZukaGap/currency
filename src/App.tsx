import 'react-native-gesture-handler';
import React from 'react';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import {RecoilRoot} from 'recoil';
import {Host} from 'react-native-portalize';

import Routes from 'navigation/Router';

export default function App() {
  return (
    <SafeAreaProvider>
      <Host>
        <RecoilRoot>
          <Routes />
        </RecoilRoot>
      </Host>
    </SafeAreaProvider>
  );
}
