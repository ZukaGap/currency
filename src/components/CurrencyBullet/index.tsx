import React from 'react';
import {Text, TouchableOpacity, View} from 'react-native';

import {CurrenciesType} from 'config/Axios/getAPI';
import {DynamicFlag} from 'primitives';

import getStyleObj from './style';

const CurrencyBullet: React.FC<CurrenciesType & {onPress?: () => void}> = ({
  name,
  code,
  rateFormated,
  diffFormated,
  diff,
  quantity,
  onPress = () => {},
}) => {
  const styles = getStyleObj();

  return (
    <TouchableOpacity
      activeOpacity={0.75}
      style={[styles.bulletContainer, styles.row]}
      onPress={onPress}>
      <DynamicFlag code={code} />
      <View style={[styles.col, {flex: 1}]}>
        <Text numberOfLines={1} style={styles.title}>
          {quantity} {name}
        </Text>
        <Text style={styles.title}>{code}</Text>
      </View>
      <View style={[styles.col]}>
        <Text style={styles.title}>{rateFormated}</Text>
        <Text style={[styles.title, diff <= 0 ? styles.green : styles.red]}>
          {diffFormated}
        </Text>
      </View>
    </TouchableOpacity>
  );
};

export default CurrencyBullet;
