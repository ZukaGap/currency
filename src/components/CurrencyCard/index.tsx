/* eslint-disable react/react-in-jsx-scope */
import {Text, View} from 'react-native';

import getStyleObj from './style';
import flags from './flags';

interface Props {
  code: string;
  quantity: number;
  rate: number;
  diff: number;
}

function CurrencyCard({code, quantity, rate, diff}: Props) {
  const styles = getStyleObj();
  const diffColor = diff >= 0 ? 'red' : 'green';
  return (
    <View style={styles.mainContainer}>
      <View style={styles.flagContainer}>
        {/* <Text>X</Text> */}
        {/* <Svg width={20} height={15}>
          <Path d={flags["AED"]} />
        </Svg> */}
        {/*  <CountryFlag countryCode={code}  */}
        {flags(code)}
      </View>
      <View style={styles.infoContainer}>
        <Text style={styles.rateText}>{`${quantity} ${code} = ${rate.toFixed(
          2,
        )} GEL`}</Text>
      </View>
      <View style={styles.diffContainer}>
        <Text style={{color: diffColor}}>{diff}</Text>
      </View>
    </View>
  );
}
export default CurrencyCard;
