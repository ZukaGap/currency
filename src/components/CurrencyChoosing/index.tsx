/* eslint-disable react/react-in-jsx-scope */
import {useState} from 'react';

import {View, Pressable} from 'react-native';

import {Dropdown} from 'react-native-element-dropdown';
import {colors} from '../../styles/colors';

import getStyleObj from './style';

import {Swap} from '../../assets/SVG';

const data = [
  {label: 'Item 1', value: '1'},
  {label: 'Item 2', value: '2'},
  {label: 'Item 3', value: '3'},
  {label: 'Item 4', value: '4'},
  {label: 'Item 5', value: '5'},
  {label: 'Item 6', value: '6'},
  {label: 'Item 7', value: '7'},
  {label: 'Item 8', value: '8'},
];

function CurrencyChoosing() {
  const styles = getStyleObj();

  const [value, setValue] = useState<any | null>(null);
  const [isFromFocus, setIsFromFocus] = useState(false);
  const [isToFocus, setIsToFocus] = useState(false);

  function swapPressHandler() {
    console.log('swap button pressed');
  }

  return (
    <View style={styles.container}>
      <Dropdown
        style={[styles.dropdown, isFromFocus && {borderColor: colors.purple}]}
        placeholderStyle={styles.placeholderStyle}
        selectedTextStyle={styles.selectedTextStyle}
        inputSearchStyle={styles.inputSearchStyle}
        data={data}
        search
        maxHeight={300}
        labelField="label"
        valueField="value"
        placeholder={!isFromFocus ? 'Convert from' : '...'}
        searchPlaceholder="Search..."
        value={value}
        onFocus={() => setIsFromFocus(true)}
        onBlur={() => setIsFromFocus(false)}
        onChange={item => {
          setValue(item.value);
          setIsFromFocus(false);
        }}
      />
      <Pressable
        style={({pressed}) => [styles.swapButton, {opacity: pressed ? 0.5 : 1}]}
        onPress={swapPressHandler}>
        <Swap width={30} height={30} fill={colors.purple} />
      </Pressable>
      <Dropdown
        style={[styles.dropdown, isToFocus && {borderColor: colors.purple}]}
        placeholderStyle={styles.placeholderStyle}
        selectedTextStyle={styles.selectedTextStyle}
        inputSearchStyle={styles.inputSearchStyle}
        data={data}
        search
        maxHeight={300}
        labelField="label"
        valueField="value"
        placeholder={!isToFocus ? 'Convert to' : '...'}
        searchPlaceholder="Search..."
        value={value}
        onFocus={() => setIsToFocus(true)}
        onBlur={() => setIsToFocus(false)}
        onChange={item => {
          setValue(item.value);
          setIsToFocus(false);
        }}
      />
    </View>
  );
}

export default CurrencyChoosing;
