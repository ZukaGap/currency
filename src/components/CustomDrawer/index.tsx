import React from 'react';
import {View, Text, TouchableOpacity} from 'react-native';
import {
  DrawerContentScrollView,
  DrawerItemList,
} from '@react-navigation/drawer';
import {useSafeAreaInsets} from 'react-native-safe-area-context';

import getStyleObj from './style';

const CustomDrawer = props => {
  const insets = useSafeAreaInsets();
  const styles = getStyleObj(insets);

  return (
    <View style={[styles.main]}>
      <DrawerContentScrollView
        {...props}
        contentContainerStyle={styles.contentContainerStyle}>
        <Text style={styles.userName}>Beka</Text>
        <DrawerItemList {...props} />
        <View style={styles.footer}>
          <TouchableOpacity onPress={() => {}} style={styles.logoutBTN}>
            <Text style={styles.logoutT}>Sign Out</Text>
          </TouchableOpacity>
        </View>
      </DrawerContentScrollView>
    </View>
  );
};

export default CustomDrawer;
