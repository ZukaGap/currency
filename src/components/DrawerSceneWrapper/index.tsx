import React from 'react';
import {StyleSheet, Platform, useWindowDimensions, View} from 'react-native';
import Animated, {interpolate, useAnimatedStyle} from 'react-native-reanimated';
import {useDrawerProgress} from '@react-navigation/drawer';
import {colors} from 'styles/colors';

const DrawerSceneWrapper = ({children}) => {
  const progress = useDrawerProgress();
  const {width} = useWindowDimensions();

  const animatedStyle = useAnimatedStyle(() => ({
    transform: [
      {perspective: 1000},
      {
        scale: interpolate(progress.value, [0, 1], [1, 0.96], 'clamp'),
      },
      {
        rotateY: `${interpolate(progress.value, [0, 1], [0, -30], 'clamp')}deg`,
      },
      {
        translateX: interpolate(
          progress.value,
          [0, 1],
          [0, Platform.OS === 'android' ? width - 130 : -60],
          'clamp',
        ),
      },
    ],
    borderRadius: interpolate(progress.value, [0, 1], [0, 20], 'clamp'),
    overflow: 'hidden',
  }));

  return (
    <Animated.View style={[styles.container, animatedStyle]}>
      {children}
    </Animated.View>
  );
};

export default DrawerSceneWrapper;

const styles = StyleSheet.create({
  container: {flex: 1},
});
