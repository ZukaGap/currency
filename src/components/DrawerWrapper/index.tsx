import React, {
  ForwardRefRenderFunction,
  forwardRef,
  useCallback,
  useImperativeHandle,
} from 'react';
import {Dimensions, StatusBar, Text, View} from 'react-native';
import {
  PanGestureHandler,
  PanGestureHandlerGestureEvent,
  TouchableOpacity,
  TouchableWithoutFeedback,
} from 'react-native-gesture-handler';
import Animated, {
  Easing,
  Extrapolate,
  Extrapolation,
  FadeInRight,
  interpolate,
  useAnimatedGestureHandler,
  useAnimatedProps,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';
import {useSafeAreaInsets} from 'react-native-safe-area-context';
import {useNavigation, useRoute} from '@react-navigation/native';

import {
  DrawerWrapperProps,
  DrawerWrapperRefProps,
} from 'types/drawerWrapperTypes';

import getStyleObj from './style';

const {width: SCREEN_WIDTH} = Dimensions.get('screen');

const THRESHOLD = SCREEN_WIDTH / 3;

// stack screens where we want navigate from Drawer
const TABS = [
  {
    key: 'homeScreen',
    name: 'Currency',
  },
  {
    key: 'fuelInfoScreen',
    name: 'Fuel',
  },
];

const DrawerWrapper: ForwardRefRenderFunction<
  DrawerWrapperRefProps,
  DrawerWrapperProps
> = ({children}, ref) => {
  const styles = getStyleObj({});
  const translateX = useSharedValue(0);
  const insets = useSafeAreaInsets();
  const route = useRoute();
  const {replace} = useNavigation();
  const animatedProps = useAnimatedProps(() => {
    const pointerEvents = translateX.value === 0 ? 'auto' : 'none';
    return {
      pointerEvents,
    };
  }, [translateX.value]);

  const panGestureEvent = useAnimatedGestureHandler<
    PanGestureHandlerGestureEvent,
    {x: number}
  >({
    onStart: (_, context) => {
      context.x = translateX.value;
    },
    onActive: (event, context) => {
      translateX.value = Math.max(event.translationX + context.x, 0);
    },
    onEnd: () => {
      if (translateX.value <= THRESHOLD) {
        translateX.value = withTiming(0);
      } else {
        translateX.value = withTiming(SCREEN_WIDTH / 1.6, {
          duration: 500,
          easing: Easing.bezier(0.25, 0.1, 0.25, 1),
        });
      }
    },
  });

  const childrenAnimStyle = useAnimatedStyle(() => ({
    transform: [
      {perspective: 100},
      {
        translateX: translateX.value,
      },
      {
        translateY: interpolate(
          translateX.value,
          [0, 100],
          [0, insets.top + 12],
          Extrapolation.CLAMP,
        ),
      },
      {
        rotate: `-${interpolate(
          translateX.value,
          [0, SCREEN_WIDTH],
          [0, 16],
          Extrapolate.CLAMP,
        )}deg`,
      },
    ],
    borderTopLeftRadius: interpolate(
      translateX.value,
      [0, 1],
      [0, 32],
      Extrapolation.CLAMP,
    ),
    paddingTop: interpolate(
      translateX.value,
      [0, SCREEN_WIDTH],
      [insets.top, 0],
      Extrapolation.CLAMP,
    ),
    overflow: 'hidden',
  }));

  const menuStyle = useAnimatedStyle(() => ({
    opacity: interpolate(translateX.value, [0, 1], [0, 1], Extrapolation.CLAMP),
    transform: [
      {
        translateY: interpolate(
          translateX.value,
          [0, 100],
          [0, insets.top],
          Extrapolation.CLAMP,
        ),
      },
    ],
    borderTopLeftRadius: interpolate(
      translateX.value,
      [0, 1],
      [0, 32],
      Extrapolation.CLAMP,
    ),
  }));

  const onPress = useCallback(() => {
    'worklet';
    if (translateX.value > 0) {
      translateX.value = withTiming(0);
    } else {
      translateX.value = withTiming(SCREEN_WIDTH / 1.6);
    }
  }, []);

  useImperativeHandle(ref, () => ({
    openDrawer: () => {
      onPress();
    },
  }));

  const closeDrawer = useCallback(() => {
    'worklet';
    if (translateX.value > 0) {
      translateX.value = withTiming(0);
    }
  }, []);

  return (
    <TouchableWithoutFeedback onPress={closeDrawer}>
      <View style={styles.container}>
        <StatusBar
          animated={true}
          translucent={true}
          backgroundColor="transparent"
        />
        <Animated.View style={[styles.drawerWrapper, menuStyle]}>
          <Text style={styles.name}>M-Economizer</Text>
          <View style={styles.drawerContent}>
            <View style={styles.screensWrapper}>
              {TABS?.map(item => (
                <TouchableOpacity
                  disabled={item?.key === route?.name}
                  onPress={() => {
                    replace(item.key);
                  }}>
                  <View
                    style={[
                      styles.screenBTN,
                      item.key === route?.name && styles.screenBTNActive,
                    ]}>
                    <Text
                      style={[
                        styles.screensTitle,
                        item?.key === route?.name && styles.screensTitleActive,
                      ]}>
                      {item?.name}
                    </Text>
                  </View>
                </TouchableOpacity>
              ))}
            </View>
          </View>
        </Animated.View>
        <PanGestureHandler
          onGestureEvent={panGestureEvent}
          failOffsetY={[-5, 5]}
          activeOffsetX={[-5, 5]}>
          <Animated.View
            entering={FadeInRight.delay(100).stiffness(50)}
            style={[styles.root, childrenAnimStyle]}>
            <Animated.View {...{animatedProps}} style={[styles.root]}>
              {children}
            </Animated.View>
          </Animated.View>
        </PanGestureHandler>
      </View>
    </TouchableWithoutFeedback>
  );
};

export default forwardRef(DrawerWrapper);
