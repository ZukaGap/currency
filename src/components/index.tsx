export {default as Header} from './Header';
export {default as CurrencyBullet} from './CurrencyBullet';
export {default as FuelPriceBullet} from './FuelPriceBullet';
export {default as FuelCompanyBullet} from './FuelCompanyBullet';
export {default as Toggle} from './Toggle';
export {default as DrawerWrapper} from './DrawerWrapper';
export {default as TabSwitchButton} from './TabSwitchButton';
