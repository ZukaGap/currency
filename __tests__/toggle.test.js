import React from 'react';
import renderer from 'react-test-renderer';

import {Toggle} from '../src/components';

describe('Toggle Button Test', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Toggle />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
