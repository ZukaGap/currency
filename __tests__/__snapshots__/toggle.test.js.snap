// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Button Test renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "marginVertical": 5,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 18,
          "fontWeight": "600",
        },
        Object {
          "color": "#FFFFFF",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "flexGrow": 1,
      }
    }
  />
  <RNCSegmentedControl
    enabled={true}
    onChange={[Function]}
    onValueChange={[Function]}
    selectedIndex={1}
    style={
      Array [
        Object {
          "height": 32,
        },
        Object {
          "marginLeft": 10,
          "width": 140,
        },
      ]
    }
    values={
      Array [
        "yes",
        "no",
      ]
    }
  />
</View>
`;
